### ISSUE TITLE
Unable to select iOS simulator for deviceName on XCode9.2 for webView tests

### ISSUE THREAD

**Issue by kool79** [appium/appium#9954]
## The problem

Appium should have ability to close all simulators **at test startup** except one which is requested in desired-capabilities (if any). Currently appium does not close previous simulator if another deviceName requested. (I use XCode 9.2 which allows parallel simulators). This  has negative effect on resource utilization and makes unable to run vebview tests (because of apple bug when we can see only one webview session - from the first device).
  Of cause, we can set:
 fullReset: true AND  resetOnSessionStartOnly: false
 but:
 * this approach does not work well (actually does not work when test session is terminated abnormally).  
 *  this significantly increases testing time (+30 sec/testcase)

## Environment

* XCode version : 9.2
* Appium version 1.7.2
* Last Appium version that did not exhibit the issue (if applicable): ?
* Desktop OS/version used to run Appium: --
* Node.js version (unless using Appium.app|exe): --
* Mobile platform/version under test: 11.2
* Real device or emulator/simulator: simulator
* Appium CLI or Appium.app|exe: 

browserName = "safari"
platformName = "iOS"
automationName = "XCUITest"
platformVersion = "11.2"

## Code To Reproduce Issue [ Good To Have ]

No code. run test with deviceName = "iPhone 7" then run test with deviceName = "iPad Air 2"
Capabilities are: 
```
browserName = "safari"
platformName = "iOS"
automationName = "XCUITest"
platformVersion = "11.2
```

**Comment by mykola-mokhnach** [comment]
> this approach does not work well (actually does not work when test session is terminated abnormally).

We can add a capability or a command to shutdown a Simulator. But we can only do it for the Simulator(s), which is/are currently under test. Giving a possibility to control other simulators might be a potential security hole. Also, I don't think it would save you as well if the session is terminated abnormally and driver.quit has not been called.

**Comment by kool79** [comment]
I prefer capability because I use appium as grid node in CI. (So I have no ability to close simulator after every terminated test)
> Giving a possibility to control other simulators might be a potential security hole. 

Why do you think it is security hole? I'm asking to close simulator devices which were opened in previous session, i.e.  by the same Simulator.app (maybe with simctl). In other words I ask to emulate behavior which XCode8 has (only one simulator process).
 
 It is the real bug for those who use appium in CI to test Safari/webViews and use xCode9. They either **will have tests which were executed on wrong devices** or all tests will be stopped after first simulator will be closed) with *Could not navigate to webview! Err: connect ECONNREFUSED ::1:27753* 
  Of cause root-cause is the apple's bug with webview but they dont care ((
 
BTW, do we have option/capability which closes simulator (without  full-reset) at the end?

**Comment by mykola-mokhnach** [comment]
> Why do you think it is security hole? I'm asking to close simulator devices which were opened in previous session, i.e. by the same Simulator.app (maybe with simctl). In other words I ask to emulate behavior which XCode8 has (only one simulator process).

Sessions are stateless from server perspective, so we don't have the information about which device was used previously by which session. Only you as a client have this info. 

However, I've just created a [PR](https://github.com/appium/appium-xcuitest-driver/pull/599) which allows access to the whole simctl interface if you have relaxed security option enabled on the server.
In order to shutdown all the running simulators, for example, you will have to run the following client code:

```java
Map <String, Object> params = new HashMap<>();
params.put("args", Lists.newArrayList("shutdown", "all"));
driver.executeScript("mobile: simctl", params);
```

This will only work if the Appium server has `--relaxed-security` argument in the command line.

**Comment by kool79** [comment]
How to execute this before the session starts? Because by doing this on active session I'll kill remote end for driver and I need to create new one. I'm on grid, so next session can be on another machine and can be queued. Also I cannot predict which devices (ipad/iphone) in which order will be requested on concrete node.

 >  we don't have the information about which device was used previously by which session
just kill them all (which marked as 'booted' in simctl)  except requested one.

**Comment by mykola-mokhnach** [comment]
you can always get the list of currently running simulators with `simctl list` and then filter by Booted status. Then you could exclude the current simulator (since you have its name/id) from this list and shutdown all the others

**Comment by kool79** [comment]
Sure, but I cannot imagine how to do this ***before*** session starts from client code, i.e. before I get instance of driver. It is impotent to kill another simulator before start, otherwise it still occupy 'vacant place' in Safari remote debugger. As you wrote #9209, we will not be able to debug even after first simulator is killed.

**Comment by mykola-mokhnach** [comment]
Ok, I got your point. will try to create a capability for that

**Comment by mykola-mokhnach** [comment]
Please check https://github.com/appium/appium-xcuitest-driver/pull/600

**Comment by RonanDrouglazet** [comment]
Hey guys ! Because I have the same problem for my own tests, i'm following this issue. So when I saw a new version for `appium-xcuitest-driver` (@2.67) with the support of `shutdownOtherSimulators` option, I've tested it as soon as possible ! But, I have an error with this new feature when enabled 

-> `[XCUITest] TypeError: allDevices.filter is not a function`

which comes from https://github.com/appium/appium-xcuitest-driver/pull/600/files#diff-bdfd0d2726c6301f92259c6fe3d4928eR100

allDevices seems to be a map of arrays {[index:string]: []}, and not directly an array (https://github.com/appium/node-simctl) :( (or did I miss something ..?)

**Comment by mykola-mokhnach** [comment]
Thanks for the report @RonanDrouglazet 
Can you please try the fix from https://github.com/appium/appium-xcuitest-driver/pull/613 ?

**Comment by kool79** [comment]
Hi, @mykola-mokhnach 
I have the same issue: `allDevices.filter is not a function` (still not try your last fix). But want to note that log contains one extra error:
`[simctl] Error: simctl error running 'terminate': An error was encountered processing the command (domain=com.apple.CoreSimulator.SimError, code=164):
Unable to lookup in current state: Shutdown`
Below is excerpt from log:

```
[Appium] Creating new XCUITestDriver (v2.67.0) session
[Appium] Capabilities:
[Appium]   appium:automationName: XCUITest
[Appium]   appium:deviceName: iPhone 7
[Appium]   appium:newCommandTimeout: 300
[Appium]   appium:platformVersion: 11.2
[Appium]   appium:resetOnSessionStartOnly: false
[Appium]   appium:shutdownOtherSimulators: true
[Appium]   browserName: safari
[Appium]   platformName: ios
[Appium]   automationName: XCUITest
[Appium]   deviceName: iPhone 7
[Appium]   newCommandTimeout: 300
[Appium]   platformVersion: 11.2
[Appium]   resetOnSessionStartOnly: false
[Appium]   shutdownOtherSimulators: true
[Appium] Applying relaxed security to XCUITestDriver as per server command line argument
..... <lines deleted>
[XCUITest] Determining device to run tests on: udid: 'B57AE78F-0AEF-4768-9588-819A0479D900', real device: false
[XCUITest] Safari test requested
[debug] [BaseDriver] Event 'appConfigured' logged at 1516845107692 (03:51:47 GMT+0200 (EET))
[debug] [BaseDriver] Event 'resetStarted' logged at 1516845107692 (03:51:47 GMT+0200 (EET))
[simctl] Error: simctl error running 'terminate': An error was encountered processing the command (domain=com.apple.CoreSimulator.SimError, code=164):
Unable to lookup in current state: Shutdown
    at Object.wrappedLogger.errorAndThrow (/usr/local/lib/node_modules/appium/node_modules/appium-support/lib/logging.js:69:13)
    at simCommand$ (/usr/local/lib/node_modules/appium/node_modules/node-simctl/lib/simctl.js:50:11)
    at tryCatch (/usr/local/lib/node_modules/appium/node_modules/babel-runtime/regenerator/runtime.js:67:40)
    at GeneratorFunctionPrototype.invoke [as _invoke] (/usr/local/lib/node_modules/appium/node_modules/babel-runtime/regenerator/runtime.js:315:22)
    at GeneratorFunctionPrototype.prototype.(anonymous function) [as throw] (/usr/local/lib/node_modules/appium/node_modules/babel-runtime/regenerator/runtime.js:100:21)
    at GeneratorFunctionPrototype.invoke (/usr/local/lib/node_modules/appium/node_modules/babel-runtime/regenerator/runtime.js:136:37)
    at <anonymous>
 Error: simctl error running 'terminate': An error was encountered processing the command (domain=com.apple.CoreSimulator.SimError, code=164):
Unable to lookup in current state: Shutdown
    at Object.wrappedLogger.errorAndThrow (/usr/local/lib/node_modules/appium/node_modules/appium-support/lib/logging.js:69:13)
    at simCommand$ (/usr/local/lib/node_modules/appium/node_modules/node-simctl/lib/simctl.js:50:11)
    at tryCatch (/usr/local/lib/node_modules/appium/node_modules/babel-runtime/regenerator/runtime.js:67:40)
    at GeneratorFunctionPrototype.invoke [as _invoke] (/usr/local/lib/node_modules/appium/node_modules/babel-runtime/regenerator/runtime.js:315:22)
    at GeneratorFunctionPrototype.prototype.(anonymous function) [as throw] (/usr/local/lib/node_modules/appium/node_modules/babel-runtime/regenerator/runtime.js:100:21)
    at GeneratorFunctionPrototype.invoke (/usr/local/lib/node_modules/appium/node_modules/babel-runtime/regenerator/runtime.js:136:37)
    at <anonymous>
[debug] [iOSSim] Cleaning mobile safari data files
[debug] [iOSSim] Checking whether simulator has been run before
[debug] [iOSSim] Simulator has not been run before
[iOSSim] Could not find Safari support directories to clean out old data. Probably there is nothing to clean out
[debug] [BaseDriver] Event 'resetComplete' logged at 1516845108061 (03:51:48 GMT+0200 (EET))
[XCUITest] Continuing without capturing logs: iOS Simulator with udid B57AE78F-0AEF-4768-9588-819A0479D900 is not running
[XCUITest] Setting up simulator
[XCUITest] TypeError: allDevices.filter is not a function
    at shutdownOtherSimulators$ (/usr/local/lib/node_modules/appium/node_modules/appium-xcuitest-driver/lib/simulator-management.js:100:41)
    at tryCatch (/usr/local/lib/node_modules/appium/node_modules/babel-runtime/regenerator/runtime.js:67:40)
    at GeneratorFunctionPrototype.invoke [as _invoke] (/usr/local/lib/node_modules/appium/node_modules/babel-runtime/regenerator/runtime.js:315:22)
    at GeneratorFunctionPrototype.prototype.(anonymous function) [as next] (/usr/local/lib/node_modules/appium/node_modules/babel-runtime/regenerator/runtime.js:100:21)
    at GeneratorFunctionPrototype.invoke (/usr/local/lib/node_modules/appium/node_modules/babel-runtime/regenerator/runtime.js:136:37)
    at <anonymous>
 TypeError: allDevices.filter is not a function
    at shutdownOtherSimulators$ (/usr/local/lib/node_modules/appium/node_modules/appium-xcuitest-driver/lib/simulator-management.js:100:41)
    at tryCatch (/usr/local/lib/node_modules/appium/node_modules/babel-runtime/regenerator/runtime.js:67:40)
    at GeneratorFunctionPrototype.invoke [as _invoke] (/usr/local/lib/node_modules/appium/node_modules/babel-runtime/regenerator/runtime.js:315:22)
    at GeneratorFunctionPrototype.prototype.(anonymous function) [as next] (/usr/local/lib/node_modules/appium/node_modules/babel-runtime/regenerator/runtime.js:100:21)
    at GeneratorFunctionPrototype.invoke (/usr/local/lib/node_modules/appium/node_modules/babel-runtime/regenerator/runtime.js:136:37)
    at <anonymous>
[debug] [XCUITest] Not clearing log files. Use `clearSystemFiles` capability to turn on.
```

**Comment by kool79** [comment]
Fix from ppium/appium-xcuitest-driver#613  works OK
